test/docker-e2e/e2e_upgrade_test.go

```diff
@@ -0,0 +1,306 @@
+}
+
+// TestCelestiaAppMinorUpgrade tests a simple upgrade from one minor version to another by swapping the binary.
+func (s *CelestiaTestSuite) TestCelestiaAppMinorUpgrade() {
+	if testing.Short() {
+		s.T().Skip("skipping celestia-app minor upgrade test in short mode")
+	}
+
+	tt := []struct {
+		Name                string
+		BaseBinaryVersion   string
+		TargetBinaryVersion string
+	}{
+		{
+			Name:                "v4.0.2-rc2 to v4.0.10",
+			BaseBinaryVersion:   "v4.0.2-rc2",
+			TargetBinaryVersion: "v4.0.10",
+		},
+		{
+			Name:                "v4.0.9-mocha to v4.0.10-mocha",
+			BaseBinaryVersion:   "v4.0.9-mocha",
+			TargetBinaryVersion: "v4.0.10-mocha",
+		},
+		{
+			Name:                "v4.0.9-arabica to v4.0.10-arabica",
+			BaseBinaryVersion:   "v4.0.9-arabica",
+			TargetBinaryVersion: "v4.0.10-arabica",
+		},
+	}
+
+	for _, tc := range tt {
+		s.Run(tc.Name, func() {
+			s.RunMinorUpgradeTest(tc.BaseBinaryVersion, tc.TargetBinaryVersion)
+		})
+	}
+}
+
+// TestCelestiaAppMajorUpgrade tests a major upgrade using the signaling mechanism.
+func (s *CelestiaTestSuite) TestCelestiaAppMajorUpgrade() {
+	if testing.Short() {
+		s.T().Skip("skipping celestia-app major upgrade test in short mode")
+	}
+
+	tt := []struct {
+		Name             string
+		BinaryVersion    string
+		TargetAppVersion uint64
+	}{
+		{
+			Name:             "v2 to v3",
+			BinaryVersion:    dockerchain.GetCelestiaTag(),
+			TargetAppVersion: 3,
+		},
+		{
+			Name:             "v3 to v4",
+			BinaryVersion:    dockerchain.GetCelestiaTag(),
+			TargetAppVersion: 4,
+		},
+	}
+
+	for _, tc := range tt {
+		s.Run(tc.Name, func() {
+			s.RunMajorUpgradeTest(tc.BinaryVersion, tc.TargetAppVersion)
+		})
+	}
+}
+
+// RunMinorUpgradeTest performs a minor version upgrade test for the celestia-app.
+// It starts a chain with the specified base version, performs a bank send transaction to verify functionality,
+// upgrades the chain to the target version, restarts it, and verifies that the chain is running the new version
+// and that bank send transactions still work.
+//
+// Example usage:
+//
+//	s.RunMinorUpgradeTest("v4.0.2-mocha", "v4.0.10-mocha")
+func (s *CelestiaTestSuite) RunMinorUpgradeTest(BaseBinaryVersion, TargetBinaryVersion string) {
+	var (
+		ctx = context.Background()
+		cfg = dockerchain.DefaultConfig(s.client, s.network).WithTag(BaseBinaryVersion)
+	)
+
+	chain, err := dockerchain.NewCelestiaChainBuilder(s.T(), cfg).Build(ctx)
+	s.Require().NoError(err)
+
+	s.T().Cleanup(func() {
+		if err := chain.Stop(ctx); err != nil {
+			s.T().Logf("Error stopping chain: %v", err)
+		}
+	})
+
+	err = chain.Start(ctx)
+	s.Require().NoError(err)
+
+	// Sanity check: Test bank send before upgrade
+	s.T().Log("Testing bank send functionality before upgrade")
+	testBankSend(s.T(), chain, cfg)
+
+	err = chain.Stop(ctx)
+	s.Require().NoError(err)
+
+	chain.UpgradeVersion(ctx, TargetBinaryVersion)
+
+	err = chain.Start(ctx)
+	s.Require().NoError(err)
```

Comment:
[blocking question]

I think this is the same question as above. we want to test for incompatibility. Stopping all nodes at the same time, then starting on new version doesn't test that.

Reply:
good point, atm only a single `Stop` at the chain level exists in the tastora API, can we create a TODO  and an issue [here](https://github.com/celestiaorg/celestia-app/issues/5220) to update the test to more correctly handle this situation?

Reply:
This is only for minor upgrade version. So if I understood it correctly, we would want to stop and start nodes in different heights instead of all at once, right?

Reply:
That's my understanding, you would want to have different nodes all on different versions and be able to incrementally upgrade individual nodes instead of all at once.